fem/transfer.cpp

```diff
@@ -871,13 +868,204 @@ void L2ProjectionGridTransfer::L2ProjectionH1Space::AllocR()
    dof_lor_dof_ho.SortRows();
    double* data = Memory<double>(dof_dofI[ndof_lor]);
 
-   R = SparseMatrix(dof_dofI, dof_dofJ, data, ndof_lor, ndof_ho,
-                    true, true, true);
-   R = 0.0;
+   SparseMatrix* R_local = new SparseMatrix(
+      dof_dofI, dof_dofJ, data, ndof_lor, ndof_ho, true, true, true);
+   (*R_local) = 0.0;
 
    dof_lor_dof_ho.LoseData();
+
+   return R_local;
+}
+
+L2ProjectionGridTransfer::SerialL2ProjectionH1Space::SerialL2ProjectionH1Space(
+   const FiniteElementSpace& fes_ho_, const FiniteElementSpace& fes_lor_)
+   : L2ProjectionH1Space(fes_ho_, fes_lor_)
+{
+   std::tie(R_sm, M_LH_sm) = ComputeSparseRAndM_LH();
+   RTxM_LH_sm = TransposeMult(*R_sm, *M_LH_sm);
+
+   R = R_sm;
+   M_LH = M_LH_sm;
+   RTxM_LH = RTxM_LH_sm;
+
+   InitializeCGSolver();
+   Ds = DSmoother(*RTxM_LH_sm);
+   pcg.SetPreconditioner(Ds);
+   pcg.SetOperator(*RTxM_LH);
+}
+
+L2ProjectionGridTransfer::SerialL2ProjectionH1Space::~SerialL2ProjectionH1Space()
+{
+   delete R_sm;
+   delete M_LH_sm;
+   delete RTxM_LH_sm;
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::LumpedMassInverse(
+   Vector& ML_inv) const
+{
+   for (int i = 0; i < ML_inv.Size(); ++i)
+   {
+      ML_inv[i] = 1.0 / ML_inv[i];
+   }
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::GetHOTDofsByVDim(
+   int vdim, const Vector& x, Vector& x_vdim_true) const
+{
+   GetDofsByVDim(fes_ho, vdim, x, x_vdim_true);
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::GetLORTDofsByVDim(
+   int vdim, const Vector& x, Vector& x_vdim_true) const
+{
+   GetDofsByVDim(fes_lor, vdim, x, x_vdim_true);
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::SetHOFromTDofsByVDim(
+   int vdim, const Vector& y_vdim_true, Vector& y) const
+{
+   SetFromDofsByVDim(fes_ho, vdim, y_vdim_true, y);
+}
+
+void L2ProjectionGridTransfer::SerialL2ProjectionH1Space::SetLORFromTDofsByVDim(
+   int vdim, const Vector& y_vdim_true, Vector& y) const
+{
+   SetFromDofsByVDim(fes_lor, vdim, y_vdim_true, y);
+}
+
+#ifdef MFEM_USE_MPI
+
+L2ProjectionGridTransfer::ParL2ProjectionH1Space::ParL2ProjectionH1Space(
+   const ParFiniteElementSpace& pfes_ho_, const ParFiniteElementSpace& pfes_lor_)
+   : L2ProjectionH1Space(pfes_ho_, pfes_lor_),
+     pfes_ho(pfes_ho_),
+     pfes_lor(pfes_lor_)
+{
+   if (pfes_ho.GetVDim() != 1)
+   {
+      pfes_ho_scalar = new ParFiniteElementSpace(pfes_ho.GetParMesh(),
+                                                 pfes_ho.FEColl(), 1);
+      pfes_lor_scalar = new ParFiniteElementSpace(pfes_lor.GetParMesh(),
+                                                  pfes_lor.FEColl(), 1);
+   }
+   else // vdim == 1
+   {
+      pfes_ho_scalar = &pfes_ho;
+      pfes_lor_scalar = &pfes_lor;
+   }
```

Comment:
Are these really necessary? In the `vdim > 1` case is there no way to work directly on the origin FE space?